packages/internal/test/marshal.test.js

```diff
@@ -0,0 +1,301 @@
+  const marshaller = makeMockMarshaller({
+    valueHook: val => {
+      unexpectedMarshallerInvocation(t, val);
+    },
+  });
+
+  const wrappedMarshaller = wrapRemoteMarshaller(marshaller);
+
+  const specimen = harden({ foo: 42, err: Error('bar') });
+
+  const capData = await wrappedMarshaller.toCapData(specimen);
+  const clone = await wrappedMarshaller.fromCapData(capData);
+
+  t.deepEqual(clone, specimen);
+});
+
+const withSlots = test.macro(async (t, { withCache } = {}) => {
+  let valueHookCalled = false;
+  const marshaller = makeMockMarshaller({
+    valueHook: val => {
+      if (valueHookCalled && withCache) {
+        unexpectedMarshallerInvocation(t, val);
+      } else {
+        valueHookCalled = true;
+        checkRemoteMarshallerValueInvariants(t, val);
+      }
+    },
+  });
+  const wrappedMarshaller = wrapRemoteMarshaller(
+    marshaller,
+    {},
+    withCache
+      ? {} // default caches
+      : {
+          slotToVal: null,
+          valToSlot: null,
+        },
+  );
+
+  const specimen = harden({ foo: 42, bar: Far('bar', { sentinel() {} }) });
+
+  // Will populate the cache
+  const capData = await wrappedMarshaller.toCapData(specimen);
+  // Will hit the cache
+  const clone = await wrappedMarshaller.fromCapData(capData);
+
+  t.deepEqual(clone, specimen);
+
+  // Will hit the cache
+  const capData2 = await wrappedMarshaller.toCapData(specimen);
+
+  const caches = {
+    slotToVal: withCache ? new Map() : null,
+    valToSlot: withCache ? new Map() : null,
+  };
+  const wrappedMarshaller2 = wrapRemoteMarshaller(marshaller, {}, caches);
+  valueHookCalled = false;
+
+  // Populate the cache
+  const clone2 = await wrappedMarshaller2.fromCapData(capData2);
+  t.true(valueHookCalled);
+  t.deepEqual(clone2, specimen);
+
+  if (withCache) {
+    const cap = specimen.bar;
+    const slot = capData2.slots[0];
+    t.is(caches.slotToVal?.get(slot), cap);
+    t.is(caches.valToSlot?.get(cap), slot);
+  }
+
+  // Hit the cache
+  const clone3 = await wrappedMarshaller2.fromCapData(capData2);
+  t.deepEqual(clone3, specimen);
+});
+
+test('wrapRemoteMarshaller - with slots - without cache', withSlots, {
+  withCache: false,
+});
+test('wrapRemoteMarshaller - with slots - with cache', withSlots, {
+  withCache: true,
+});
+
+test('wrapRemoteMarshaller - read-only marshaller', async t => {
+  const marshaller = makeMockMarshaller({
+    valueHook: (val, direction) => {
+      if (direction === 'from') {
+        unexpectedMarshallerInvocation(t, val);
+      } else {
+        checkRemoteMarshallerValueInvariants(t, val);
+      }
+    },
+    readOnly: true,
+  });
+  const wrappedMarshaller = wrapRemoteMarshaller(marshaller);
+
+  const specimen = harden({ foo: 42, bar: Far('bar', { sentinel() {} }) });
+
+  const capData = await wrappedMarshaller.toCapData(specimen);
+  const clone = await wrappedMarshaller.fromCapData(capData);
+
+  t.deepEqual(clone, { ...specimen, bar: makeInaccessibleVal('bar') });
```

Comment:
for the read-only marshaller, correctness is about establishing that `toCapData` maps non-published values to `null` slots, right? Does this test do that somehow?